### PR TITLE
Spid rule update

### PR DESCRIPTION
**Pull request by noodlemoodle**
- update setup spid to put directly to bare ucjdao
- add predicate to check if ucj is user serviceprovider junction

### CODE REVIEW

File: src/foam/nanos/crunch/predicate/IsUserServiceProviderJunction.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+ * @license
+ * Copyright 2021 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+ foam.CLASS({
+  package: 'foam.nanos.crunch.predicate',
+  name: 'IsUserServiceProviderJunction',
+  extends: 'foam.mlang.predicate.AbstractPredicate',
+
```
Comment (jlhughes) [inline src/foam/nanos/crunch/predicate/IsUserServiceProviderJunction.js:15]: documentation please.

File: src/foam/nanos/auth/ServiceProvider.js [hunk 2]
```diff
@@ -97,11 +98,16 @@ foam.CLASS({
         try {
           // for each capability in the grantPath of the spid capability,
           // find the ucj and update its status to granted, or create a ucj none found
-
           UserCapabilityJunction ucj;
-          Subject subject = new Subject(user);
           for ( Capability capability : grantPath ) {
-            ucj = crunchService.updateUserJunction(x, subject, capability.getId(), null, CapabilityJunctionStatus.GRANTED);
+            if ( ! includeSpid && capability.getId() == getId() ) continue;
```
Comment (jlhughes) [inline src/foam/nanos/auth/ServiceProvider.js:107]: Please document includeSpid.

File: src/foam/nanos/auth/ServiceProvider.js [hunk 3]
```diff
@@ -80,7 +80,12 @@ foam.CLASS({
       name: 'setupSpid',
       args: [
         { name: 'x', javaType: 'foam.core.X' },
-        { name: 'user', javaType: 'foam.nanos.auth.User' }
+        { name: 'user', javaType: 'foam.nanos.auth.User' },
+        {
+          name: 'includeSpid',
+          javaType: 'Boolean',
+          documentation: 'If true, include the top level spid capability when granting the capability hierarchy'
+        }
```
Comment (zahnno) [inline src/foam/nanos/auth/ServiceProvider.js:88]: What does this solve?

File: src/foam/nanos/crunch/rules.jrl [hunk 4]
```diff
@@ -21,8 +21,9 @@ p({
   "documentation":"Re-assign user spid on CREATE of User-ServiceProvider junction",
   "daoKey":"userCapabilityJunctionDAO",
   "operation":0,
-  "after":true,
+  "after":false,
   "enabled":true,
+  "predicate":{ "class":"foam.nanos.crunch.predicate.IsUserServiceProviderJunction" },
   "action":{
     "class":"foam.nanos.auth.SetUserServiceProviderJunctionRuleAction"
   },
```
Comment (zahnno) [inline src/foam/nanos/crunch/rules.jrl:29]: I don't see a scenario where we'll be "switching" a user's spid by adding a user capability junction. Most cases we set a new spid value on an object. We have another rule which checks for this spid update on user and handles the setup of the spid capabilities. I don't think this rule is necessary.

Comment (noodlemoodle) [reply src/foam/nanos/crunch/rules.jrl:29]: i think thats fine if we don't need to cover this scenario, but then we might want prevent puts to "userCapabilityJunctionDAO" for user-serviceprovider junctions altogether

File: src/foam/nanos/auth/SetUserServiceProviderJunctionRuleAction.js [hunk 5]
```diff
@@ -73,7 +73,7 @@ foam.CLASS({
             serviceProvider.removeSpid(x, user);
 
             // get grantpath and create/reput grantpath
-            serviceProvider.setupSpid(x, user);
+            serviceProvider.setupSpid(x, user, false);
```
Comment (lchanmann) [inline src/foam/nanos/auth/SetUserServiceProviderJunctionRuleAction.js:76]: Maybe can check if the user.spid is same as the service provider before re-setting up the user spid?

Comment (jlhughes) [reply src/foam/nanos/auth/SetUserServiceProviderJunctionRuleAction.js:76]: Yes, looking at performance, this seems to be an unnecessary re-assignment when the spid has not changed. 
As @lchanmann mentioned, can we do something like:
`            User user = (User) ucj.findSourceId(x).fclone();
            ServiceProvider serviceProvider = (ServiceProvider) target;
            if ( user.getSpid() == serviceProvider.getId() ) {
              return;
            }
`
at line 53?

**Pull request closed without merging**